### PR TITLE
vtctl,vindexes: logs warnings and export stat for unknown vindex params

### PR DESCRIPTION
**Pull request by maxenglander**
## Description

Surface unknown Vindex params:

 * As a VTGate gauge `VindexUnknownParams`
 * As a row in VTGate vschema stats
 * As warnings in Vtctl `ApplyVSchema` output

There are probably additional places where this could be surfaced, but this seems like a good starting point.

Additionally:

 * `ApplyVSchema --dry-run` will now validate the VSchema.
 * [715633c](https://github.com/vitessio/vitess/pull/13322/commits/715633cc60ad3271acec064805f7e1745ba5d297) addresses feedback from @GuptaManan100 that I missed in https://github.com/vitessio/vitess/pull/12951#discussion_r1227614003

## Related Issue(s)

 * Follow up to https://github.com/vitessio/vitess/pull/12951
 * Addresses https://github.com/vitessio/vitess/issues/12476

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was [added](https://github.com/vitessio/website/pull/1501)

### CODE REVIEW

**Comment by maxenglander** [comment]
Flagging this

>  [ ] vtctl command output order should be stable and awk-able.

I think right now I'm not satisfying this stability point. Will be a bit awkward to get the the warnings to show at the end of the command output since there are other conditions that will currently exit the command early, but can do this if it's important.

File: changelog/18.0/18.0.0/summary.md [hunk 1]
```diff
@@ -30,4 +32,10 @@ The `k8stopo` has been deprecated in Vitess 17, also see https://github.com/vite
...
 
-The `vtgr` has been deprecated in Vitess 17, also see https://github.com/vitessio/vitess/issues/13300. With Vitess 18 `vtgr` has been removed.
\ No newline at end of file
+The `vtgr` has been deprecated in Vitess 17, also see https://github.com/vitessio/vitess/issues/13300. With Vitess 18 `vtgr` has been removed.
+
+### <a id="new-stats"/>New stats
+
+#### <a id="vtgate-vschema-vindex-unknown-params"/>VTGate VSchema Vindex unknown params
+
+Gauges unknown Vindex params found in the latest VSchema pulled from the topology.
```
Comment (harshit-gangal) [inline changelog/18.0/18.0.0/summary.md:41]: should provide the metric name here.

File: go/vt/topo/vschema.go [hunk 2]
```diff
@@ -53,6 +49,17 @@ func (ts *Server) SaveVSchema(ctx context.Context, keyspace string, vschema *vsc
 	return err
 }
 
+// ValidateAndSaveVSchema first validates the VSchema, then saves it.
+// If the VSchema is empty, just remove it.
+func (ts *Server) ValidateAndSaveVSchema(ctx context.Context, keyspace string, vschema *vschemapb.Keyspace) error {
+	_, err := vindexes.BuildKeyspace(vschema)
+	if err != nil {
```
Comment (harshit-gangal) [inline go/vt/topo/vschema.go:56]: can not reuse the existing `SaveVSchema` and extend it to also validate the Vschema?

File: go/vt/vtctl/vtctl_test.go [hunk 3]
```diff
@@ -32,6 +33,54 @@ import (
...
+		},
+		{
+			name: "UnknownParamsLoggedWithDryRun",
+			args: []string{"--vschema", "{\"sharded\":true,\"vindexes\":{\"hash_vdx\":{\"type\":\"hash\",\"params\":{\"foo\":\"bar\",\"hello\":\"world\"}},\"binary_vdx\":{\"type\":\"binary\",\"params\":{\"hello\":\"world\"}}}}", "--dry-run", ks},
+			want: "/New VSchema object:\n{\n  \"sharded\": true,\n  \"vindexes\": {\n    \"binary_vdx\": {\n      \"type\": \"binary\",\n      \"params\": {\n        \"hello\": \"world\"\n      }\n    },\n    \"hash_vdx\": {\n      \"type\": \"hash\",\n      \"params\": {\n        \"foo\": \"bar\",\n        \"hello\": \"world\"\n      }\n    }\n  }\n}\nIf this is not what you expected, check the input data \\(as JSON parsing will skip unexpected fields\\)\\.\n\n.*W0614 .* vtctl.go:.* Unknown param in vindex hash_vdx: foo\nW0614 .* vtctl.go:.* Unknown param in vindex hash_vdx: hello\nW0614 .* vtctl.go:.* Unknown param in vindex binary_vdx: hello\nDry run: Skipping update of VSchema",
```
Comment (harshit-gangal) [inline go/vt/vtctl/vtctl_test.go:65]: the test is hard to read, can the vschema be loaded from a file?

File: go/vt/vtgate/vtgate.go [hunk 4]
```diff
@@ -179,6 +179,8 @@ var (
 	// catch the initial load stats.
 	vschemaCounters = stats.NewCountersWithSingleLabel("VtgateVSchemaCounts", "Vtgate vschema counts", "changes")
 
+	vschemaVindexUnknownParams = stats.NewGauge("VSchemaVindexUnknownParams", "Number of params unrecognized by VSchema Vindexes")
```
Comment (harshit-gangal) [inline go/vt/vtgate/vtgate.go:182]: nit: `stats.NewGauge("VindexUnknownParameters", "Number of parameters unrecognized by Vindexes")`

File: go/vt/topo/vschema.go (hunk 2 above)
Comment (maxenglander) [reply go/vt/topo/vschema.go:56]: I will revert `SaveVSchema` to what it was before. 

The idea behind splitting out validation from saving was to enable `ApplyVSchema` to validate when `--dry-run` is enabled without having `vindexes.BuildKeyspace` called multiple times.

File: go/vt/vtgate/vtgate.go [hunk 5]
```diff
@@ -186,6 +186,8 @@ var (
 
 	vstreamSkewDelayCount = stats.NewCounter("VStreamEventsDelayedBySkewAlignment",
 		"Number of events that had to wait because the skew across shards was too high")
+
+	vindexUnknownParams = stats.NewGauge("VindexUnknownParams", "Number of params unrecognized by Vindexes")
```
Comment (harshit-gangal) [inline go/vt/vtgate/vtgate.go:190]: kindly change the metric name to `VindexUnknownParameters`

Comment (harshit-gangal) [reply go/vt/vtgate/vtgate.go:190]: Similar change would be needed in the release notes.

**Review by harshit-gangal — APPROVED** [review]
just one last comment. rest looks good.

File: go/vt/vtgate/vtgate.go (hunk 5 above)
Comment (maxenglander) [reply go/vt/vtgate/vtgate.go]: Done! Updated website PR too

**Pull request merged**